kolibri/core/assets/src/kolibri_app.js

```diff
@@ -68,6 +72,72 @@ export default class KolibriApp extends KolibriModule {
       store.registerModule(name, module);
     });
 
+    if (process.env.NODE_ENV !== 'production') {
+      /* eslint-disable no-inner-declarations */
+      // Register any schemas we have defined for vuex state
+      function registerSchema(schema, moduleName, subPaths = []) {
+        forEach(schema, (subSchema, propertyName) => {
+          // Must be a plain object to be a valid schema spec
+          // And have at least a default key, and one of type or validator
+          logging.debug(subSchema);
+          if (isPlainObject(subSchema)) {
+            if (
+              typeof subSchema.default !== 'undefined' &&
+              (subSchema.type || (subSchema.validator && subSchema.validator instanceof Function))
+            ) {
+              function getter(state) {
+                if (moduleName) {
+                  state = state[moduleName];
+                }
+                subPaths.forEach(path => {
+                  state = state[path];
+                });
+                return state[propertyName];
+              }
+              function callback(newValue) {
+                let fail = false;
+                if (subSchema.type) {
+                  if (subSchema.type === Object) {
+                    if (!isPlainObject(newValue)) {
+                      fail = true;
+                    }
+                  } else {
+                    fail = !(newValue instanceof subSchema.type);
+                  }
+                }
+                if (subSchema.validator) {
+                  if (!subSchema.validator(newValue)) {
+                    fail = true;
+                  }
+                }
+                if (fail) {
+                  logging.error(
+                    `Validation failed for property: ${[...subPaths, propertyName]} in module ${
+                      moduleName ? moduleName : 'root'
+                    }`
+                  );
```

Comment:
similar feedback I make occasionally: recommend `return`ing early rather than "falling through" the conditionals and checking result at the end. For example:

```javascript
function callback(newValue) {
  const prop = [...subPaths, propertyName];
  const mod = moduleName ? moduleName : 'root';
  if (subSchema.type) {
    if (subSchema.type === Object && !isPlainObject(newValue)) {
      logging.error(
        `Validation failed for property: ${property} in module ${mod}: ${newValue} should be Object`
      );
      return;
    } else if (!(newValue instanceof subSchema.type)) {
      logging.error(
        `Validation failed for property: ${property} in module ${mod}: ${newValue} should be ${subSchema.type}`
      );
      return;
    }
  }
  if (subSchema.validator && !subSchema.validator(newValue)) {
    logging.error(
      `Validation failed for property: ${property} in module ${mod}: ${newValue} failed validation.`
    );
  }
}
```

two benefits:

1. makes explicit when processing is 'complete'
2. individually tailored messages for each case